x/outline-connectivity/main.go

```diff
@@ -255,30 +143,38 @@ func main() {
 			resolverAddress := net.JoinHostPort(resolverHost, "53")
 			for _, proto := range strings.Split(*protoFlag, ",") {
 				proto = strings.TrimSpace(proto)
-				// var dnsClient := dns.Client{Net: proto}
-				var dnsDial DNSDial
-				if proto == "tcp" {
-					dnsDial, err = makeTCPDialer(proxyAddress, config.CryptoKey, config.Prefix)
-				} else {
-					dnsDial, err = makeUDPDialer(proxyAddress, config.CryptoKey)
-				}
-				if err != nil {
-					log.Fatalf("Failed to create DNS resolver: %#v", err)
-				}
+
 				testTime := time.Now()
-				testErr := testResolver(dnsDial, resolverAddress, *domainFlag)
+				var testErr error
+				var testDuration time.Duration
+				switch proto {
+				case "tcp":
+					dialer, err := makeStreamDialer(proxyAddress, config.CryptoKey, config.Prefix)
+					if err != nil {
+						log.Fatalf("Failed to create StreamDialer: %v", err)
+					}
+					testDuration, testErr = connectivity.TestStreamDialerConnectivity(context.Background(), dialer, resolverAddress, *domainFlag)
+				case "udp":
+					listener, err := makePacketListener(proxyAddress, config.CryptoKey)
+					if err != nil {
+						log.Fatalf("Failed to create PacketListener: %v", err)
+					}
+					dialer := transport.PacketListenerDialer{Listener: listener}
+					testDuration, testErr = connectivity.TestPacketDialerConnectivity(context.Background(), dialer, resolverAddress, *domainFlag)
+				default:
+					log.Fatalf(`Invalid proto %v. Must be "tcp" or "udp"`, proto)
+				}
 				debugLog.Printf("Test error: %v", testErr)
-				duration := time.Since(testTime)
 				if testErr == nil {
 					success = true
 				}
 				record := jsonRecord{
-					Time:       testTime.UTC().Truncate(time.Second),
-					DurationMs: duration.Milliseconds(),
 					Proxy:      proxyAddress,
 					Resolver:   resolverAddress,
 					Proto:      proto,
 					Prefix:     config.Prefix.String(),
+					Time:       testTime.UTC().Truncate(time.Second),
+					DurationMs: testDuration.Milliseconds(),
```

Comment:
We may also use ISO8601 duration format: https://en.wikipedia.org/wiki/ISO_8601#Durations